compiler/rustc_llvm/llvm-wrapper/RustWrapper.cpp

```diff
@@ -203,29 +203,45 @@ static Attribute::AttrKind fromRust(LLVMRustAttribute Kind) {
   report_fatal_error("bad AttributeKind");
 }
 
+template<typename T> static inline void AddAttribute(T *t, unsigned Index, Attribute Attr) {
+#if LLVM_VERSION_LT(14, 0)
+  t->addAttribute(Index, Attr);
+#else
+  t->addAttributeAtIndex(Index, Attr);
+#endif
+}
+
 extern "C" void LLVMRustAddCallSiteAttribute(LLVMValueRef Instr, unsigned Index,
                                              LLVMRustAttribute RustAttr) {
   CallBase *Call = unwrap<CallBase>(Instr);
   Attribute Attr = Attribute::get(Call->getContext(), fromRust(RustAttr));
-  Call->addAttribute(Index, Attr);
+  AddAttribute(Call, Index, Attr);
 }
 
 extern "C" void LLVMRustAddCallSiteAttrString(LLVMValueRef Instr, unsigned Index,
                                               const char *Name) {
   CallBase *Call = unwrap<CallBase>(Instr);
   Attribute Attr = Attribute::get(Call->getContext(), Name);
-  Call->addAttribute(Index, Attr);
+  AddAttribute(Call, Index, Attr);
 }
 
+static inline void AddCallAttributes(CallBase *Call, unsigned Index, const AttrBuilder& B) {
+  AttributeList Attrs = Call->getAttributes();
+#if LLVM_VERSION_LT(14, 0)
+  Attrs = Attrs.addAttributes(Call->getContext(), Index, B);
+#else
+  Attrs = Attrs.addAttributesAtIndex(Call->getContext(), Index, B);
+#endif
+  Call->setAttributes(Attrs);
+}
 
 extern "C" void LLVMRustAddAlignmentCallSiteAttr(LLVMValueRef Instr,
                                                  unsigned Index,
                                                  uint32_t Bytes) {
   CallBase *Call = unwrap<CallBase>(Instr);
   AttrBuilder B;
   B.addAlignmentAttr(Bytes);
-  Call->setAttributes(Call->getAttributes().addAttributes(
-      Call->getContext(), Index, B));
+  AddCallAttributes(Call, Index, B);
```

Comment:
I realize this is kind of a pre-existing problem, but why does this need a separate AddCallAttributes? Couldn't it use exactly the same code as LLVMRustAddAlignmentAttr? That is `AddAttribute(Call, Index, Attribute::getWithAlignment(Call->getContext(), llvm::Align(Bytes)));` or so? Same for the other uses of AddCallAttributes() below.

Reply:
It's because these functions are getting a uint{32,64}_t in instead of an LLVMRustAttribute or similar, so we have to bounce through the AttrBuilder to go from the uint* to the Attribute. I've felt like there was probably room for cleanup both times I've been in here, but haven't had time to go wading around.

Reply:
Right, but this applies the same way to the corresponding setters on `Function` (LLVMRustAddAlignmentAttr etc). Those were changed to use `Attribute::getXXX()` methods in #88289, but the same wasn't done for the call-site functions.

Reply:
Done.